File: backend/app/api/area_board.py
"""Area Board endpoint (Phase 11 — GUI_DESIGN §6; PROJECT_PROFILE §21).

``GET /area-board`` — one read for the whole Management view
(`app.application.area_board`): every active Area of the Department
with the SHARED Area monitoring shape (`app.api.area_inventory` — the
same contract `GET /areas/{id}/inventory` answers the Scan Station
with), its active Operations, the scrapped quantity per PN, and, for a
terminal Area, the stocked lines with their active allocation.

The All Areas overview and the per-Area detail are two presentations of
this one answer, so switching tabs never re-reads and the two can never
show different quantities. ``department_id`` selects the Department;
omitted, the single active Department is used — none is 404, several is
409 naming them.

A read: nothing is written, and every derived time value (the dwell
time, the due countdown) is left to the view's shared clock — the
response carries only the fixed source timestamps and dates.
"""

from fastapi import APIRouter
from pydantic import BaseModel

from app.api.area_inventory import (
    AreaInventoryResponse,
    OperationRef,
    area_inventory_response,
    operation_ref,
)
from app.api.dependencies import SessionDep
from app.application import area_board

router = APIRouter(prefix="/api")


class StockedLineResponse(BaseModel):
    """One PN held in a terminal Area, with its active allocation."""

    part_number: str
    quantity: int
    allocated_quantity: int


class ScrappedLineResponse(BaseModel):
    part_number: str
    quantity: int


class AreaBoardAreaResponse(BaseModel):
    # The shared Area monitoring model, identical to the Scan Station's.
    inventory: AreaInventoryResponse
    operations: list[OperationRef]
    scrapped: list[ScrappedLineResponse]
    # Terminal Areas only (Stockroom): manufacturing-complete quantity
    # whose flows are closed, so it is never part of the inventory.
    stocked: list[StockedLineResponse]


class AreaBoardDepartmentRef(BaseModel):
    id: int
    name: str


class AreaBoardResponse(BaseModel):
    department: AreaBoardDepartmentRef
    areas: list[AreaBoardAreaResponse]


def _area(entry: area_board.AreaBoardArea) -> AreaBoardAreaResponse:
    return AreaBoardAreaResponse(
        inventory=area_inventory_response(entry.inventory),
        operations=[operation_ref(operation) for operation in entry.operations],
        scrapped=[
            ScrappedLineResponse(part_number=part_number, quantity=quantity)
            for part_number, quantity in sorted(entry.scrapped.items())
        ],
        stocked=[
            StockedLineResponse(
                part_number=line.part_number,
                quantity=line.quantity,
                allocated_quantity=line.allocated_quantity,
            )
            for line in entry.stocked
        ],
    )


@router.get("/area-board")
def get_area_board(session: SessionDep, department_id: int | None = None) -> AreaBoardResponse:
    board = area_board.area_board(session, department_id)
    return AreaBoardResponse(
        department=AreaBoardDepartmentRef(id=board.department.id, name=board.department.name),
        areas=[_area(entry) for entry in board.areas],
    )

File: frontend/src/api/area-board.ts
// Area Board read model API (Phase 11 — GUI_DESIGN §6;
// PROJECT_PROFILE §21).
//
// One read: `GET /api/area-board` returns every active Area of the
// Department with the SHARED Area monitoring model (./area-inventory —
// the same contract the Scan Station reads, the PN's open demand
// context and the scrapped quantity per PN included), its active
// Operations, and, for the terminal Stockroom, the stocked lines with
// the PN's active allocation and its OPEN demand context.
//
// The All Areas overview and the per-Area detail are two presentations
// of this one answer: switching tabs never re-reads, and the two modes
// can never show different quantities. The server sends no derived time
// value — dwell times and due countdowns derive at render from the
// fixed timestamps and dates plus the shared UI clock (§3.12).
//
// Production-safe: no mock data, no framework imports.

import type {
  AreaInventory,
  AreaInventoryWire,
  DemandContext,
  DemandContextWire,
  OperationRef,
} from './area-inventory';
import {
  toAreaInventory,
  toDemandContext,
  toOperationRef,
} from './area-inventory';
import { apiRequest } from './client';

/** One PN held in a terminal Area (the Stockroom column / group). */
export interface StockedLine {
  partNumber: string;
  /** Gross stocked quantity of the PN in THIS Area. */
  quantity: number;
  /** The PN's ACTIVE allocation — a PN-level connection to demand,
   * never divided per Area, so the row states both values. */
  allocatedQuantity: number;
  /** The PN's OPEN demands in the canonical order — the same
   * monitoring context an inventory row carries, so a stocked PN still
   * worked FOR an open demand names it (empty once every Work Order of
   * the PN is complete). */
  demands: DemandContext[];
}

export interface AreaBoardArea {
  /** The shared Area monitoring model, identical to the station's. */
  inventory: AreaInventory;
  operations: OperationRef[];
  /** Terminal Areas only; empty everywhere else. */
  stocked: StockedLine[];
}

export interface AreaBoard {
  department: { id: number; name: string };
  areas: AreaBoardArea[];
}

interface AreaBoardAreaWire {
  inventory: AreaInventoryWire;
  operations: {
    id: number;
    code: string;
    name: string | null;
    is_external: boolean;
  }[];
  stocked: {
    part_number: string;
    quantity: number;
    allocated_quantity: number;
    demands: DemandContextWire[];
  }[];
}

interface AreaBoardWire {
  department: { id: number; name: string };
  areas: AreaBoardAreaWire[];
}

function toArea(wire: AreaBoardAreaWire): AreaBoardArea {
  return {
    inventory: toAreaInventory(wire.inventory),
    operations: wire.operations.map(toOperationRef),
    stocked: wire.stocked.map((line) => ({
      partNumber: line.part_number,
      quantity: line.quantity,
      allocatedQuantity: line.allocated_quantity,
      demands: line.demands.map(toDemandContext),
    })),
  };
}

/**
 * Load the board. `departmentId` selects the Department; omitted, the
 * server resolves the single active Department and refuses an ambiguous
 * configuration (several active Departments) with an explicit message
 * naming them.
 */
export async function loadAreaBoard(
  departmentId: number | null,
): Promise<AreaBoard> {
  const query =
    departmentId !== null
      ? `?department_id=${encodeURIComponent(departmentId)}`
      : '';
  const wire = await apiRequest<AreaBoardWire>(`/api/area-board${query}`);
  return {
    department: wire.department,
    areas: wire.areas.map(toArea),
  };
}
